File: java/src/main/java/io/github/securekit/CsrfTokenManager.java
package io.github.securekit;

import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.util.Base64;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;

/**
 * Stateless CSRF token issuance/verification bound to a session ID and a
 * server-side secret, using HMAC-SHA256.
 */
public final class CsrfTokenManager {

    private final byte[] secret;
    private final long ttlMillis;

    /**
     * @param secret    at least 32 random bytes, kept server-side
     * @param ttlMillis token lifetime in ms; &lt;=0 disables expiry
     */
    public CsrfTokenManager(byte[] secret, long ttlMillis) {
        this.secret = secret.clone();
        this.ttlMillis = ttlMillis;
    }

    public String generate(String sessionId) {
        byte[] nonce = SecureRandomUtil.bytes(16);
        long issuedAt = System.currentTimeMillis();
        byte[] tag = sign(sessionId, nonce, issuedAt);

        ByteBuffer buf = ByteBuffer.allocate(8 + nonce.length + tag.length);
        buf.putLong(issuedAt).put(nonce).put(tag);
        return Base64.getUrlEncoder().withoutPadding().encodeToString(buf.array());
    }

    public boolean verify(String sessionId, String token) {
        byte[] raw;
        try {
            raw = Base64.getUrlDecoder().decode(token);
        } catch (IllegalArgumentException e) {
            return false;
        }
        if (raw.length < 8 + 16 + 32) {
            return false;
        }
        ByteBuffer buf = ByteBuffer.wrap(raw);
        long issuedAt = buf.getLong();
        byte[] nonce = new byte[16];
        buf.get(nonce);
        byte[] tag = new byte[32];
        buf.get(tag);

        if (ttlMillis > 0 && System.currentTimeMillis() - issuedAt > ttlMillis) {
            return false;
        }
        byte[] expected = sign(sessionId, nonce, issuedAt);
        return ConstantTime.equals(expected, tag);
    }

    private byte[] sign(String sessionId, byte[] nonce, long issuedAt) {
        try {
            Mac mac = Mac.getInstance("HmacSHA256");
            mac.init(new SecretKeySpec(secret, "HmacSHA256"));
            mac.update(sessionId.getBytes(StandardCharsets.UTF_8));
            mac.update(nonce);
            mac.update(ByteBuffer.allocate(8).putLong(issuedAt).array());
            return mac.doFinal();
        } catch (NoSuchAlgorithmException | InvalidKeyException e) {
            throw new IllegalStateException("securekit: HMAC-SHA256 unavailable", e);
        }
    }
}

File: javascript/src/csrfTokenManager.js
'use strict';

const crypto = require('crypto');
const { secureRandomBytes } = require('./secureRandom');
const { constantTimeEqual } = require('./constantTime');

/**
 * Stateless CSRF token issuance/verification bound to a session ID and a
 * server-side secret, using HMAC-SHA256.
 */
class CsrfTokenManager {
  /**
   * @param {Buffer} secret at least 32 random bytes, kept server-side
   * @param {number} [ttlMs=3600000] token lifetime in ms; <=0 disables expiry
   */
  constructor(secret, ttlMs = 3600000) {
    this.secret = secret;
    this.ttlMs = ttlMs;
  }

  generate(sessionId) {
    const nonce = secureRandomBytes(16);
    const issuedAt = Date.now();
    const tag = this._sign(sessionId, nonce, issuedAt);
    const issuedAtBuf = Buffer.alloc(8);
    issuedAtBuf.writeBigUInt64BE(BigInt(issuedAt));
    return Buffer.concat([issuedAtBuf, nonce, tag]).toString('base64url');
  }

  verify(sessionId, token) {
    let raw;
    try {
      raw = Buffer.from(token, 'base64url');
    } catch {
      return false;
    }
    if (raw.length < 8 + 16 + 32) return false;

    const issuedAt = Number(raw.readBigUInt64BE(0));
    const nonce = raw.subarray(8, 24);
    const tag = raw.subarray(24, 56);

    if (this.ttlMs > 0 && Date.now() - issuedAt > this.ttlMs) {
      return false;
    }
    const expected = this._sign(sessionId, nonce, issuedAt);
    return constantTimeEqual(expected, tag);
  }

  _sign(sessionId, nonce, issuedAt) {
    const issuedAtBuf = Buffer.alloc(8);
    issuedAtBuf.writeBigUInt64BE(BigInt(issuedAt));
    const h = crypto.createHmac('sha256', this.secret);
    h.update(Buffer.concat([Buffer.from(sessionId, 'utf8'), nonce, issuedAtBuf]));
    return h.digest();
  }
}

module.exports = { CsrfTokenManager };
